### PR TITLE
executor: Change makeJoinRowToChunk to account for virtual rows

### LINKED ISSUE #5771: join has not considered the virtual rows.
Please answer these questions before submitting your issue. Thanks!

1. What did you do?
The SQL statement to generate sequence:

```
select @NUM := @NUM + 1 as NUM from (
select 1 union all select 2 union all select 3 union all select 4 union all select 5 union all
select 6 union all select 7 union all select 8 union all select 9 union all select 10
 ) a inner join (
select 1 union all select 2 union all select 3 union all select 4 union all select 5 union all
select 6 union all select 7 union all select 8 union all select 9 union all select 10
 ) b, (SELECT @NUM := 0) d;
```

2. What did you expect to see?
MySQL:
+------+
| NUM  |
+------+
|    1 |
|    2 |
.....
|   99 |
|  100 |
+------+
100 rows in set (0.08 sec)

MySQL version: Server version: 5.7.21 MySQL Community Server (GPL)

3. What did you see instead?
Empty set (0.00 sec)

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
Release Version: v1.1.0-alpha.1-53-g1f1e5f5
Git Commit Hash: 1f1e5f53eb66acb9ef534d87f4e8daeae2552f73
Git Commit Branch: master
UTC Build Time:  2018-02-01 03:33:14
GoVersion:  go version go1.9.2 linux/amd64

### PR DESCRIPTION
**Pull request by MC-Squared**
makeJoinRowToChunk was calling AppendPartialRow twice, which does not increment the virtual rows.

I'm not 100% on the implications of this change, but it hasn't broken any tests + fixes the examples described in the issue.

fix #5771

### CODE REVIEW

File: executor/join_test.go [hunk 1]
```diff
@@ -215,6 +215,24 @@ func (s *testSuite) TestJoin(c *C) {
...
+			SELECT 3
+		) a
+		INNER JOIN
+		( SELECT 1 UNION ALL
+			SELECT 2 UNION ALL
+			SELECT 3
+		) b,
+		(SELECT @NUM := 0) d;
+	`)
+	result.Check(testkit.Rows("1", "2", "3", "4", "5", "6", "7", "8", "9"))
```
Comment (zz-jason) [inline executor/join_test.go]: another simpler test cast:
```sql
select 1 from (select 1) t1, (select 1) t2;
``

Comment (MC-Squared) [reply executor/join_test.go]: That test is included above at lines 219-220. Perhaps should be separated by a newline to indicate two separate test cases.

Comment (MC-Squared) [reply executor/join_test.go]: That test is included above at lines 219-220. Perhaps should be separated by a newline to indicate two separate test cases.

File: executor/join_result_generators.go [hunk 2]
```diff
@@ -122,7 +122,7 @@ func (outputer *baseJoinResultGenerator) makeJoinRowToBuffer(buffer []types.Datu
 }
 
 func (outputer *baseJoinResultGenerator) makeJoinRowToChunk(chk *chunk.Chunk, lhs, rhs chunk.Row) {
-	chk.AppendPartialRow(0, lhs)
+	chk.AppendRow(lhs)
```
Comment (zz-jason) [inline executor/join_result_generators.go]: how about adding the github issue link here as a comment?

**Comment by zz-jason** [comment]
@MC-Squared merge join has the same problem, maybe you can have a try.

**Comment by MC-Squared** [comment]
@zz-jason I'm not that familiar with merge joins, could you provide an SQL query that shows the problem?

**Comment by zz-jason** [comment]
@MC-Squared never mind, this PR can also fix sort merge join.

File: executor/join_result_generators.go [hunk 3]
```diff
@@ -122,7 +122,9 @@ func (outputer *baseJoinResultGenerator) makeJoinRowToBuffer(buffer []types.Datu
 }
 
 func (outputer *baseJoinResultGenerator) makeJoinRowToChunk(chk *chunk.Chunk, lhs, rhs chunk.Row) {
-	chk.AppendPartialRow(0, lhs)
+	// Call AppendRow() first to increment the virtual rows
```
Comment (zz-jason) [inline executor/join_result_generators.go]: add a dot `.` in the end of line.

File: executor/join_result_generators.go [hunk 4]
```diff
@@ -122,7 +122,9 @@ func (outputer *baseJoinResultGenerator) makeJoinRowToBuffer(buffer []types.Datu
...
+	// See: https://github.com/pingcap/tidb/issues/5771
```
Comment (zz-jason) [inline executor/join_result_generators.go]: s/See/Fix/

**Comment by XuHuaiyu** [comment]
LGTM

**Comment by coocood** [comment]
@MC-Squared 
Thank you for your contribution!

**Pull request merged**